### ISSUE TITLE
BDD Kernel & Factory usage

### ISSUE THREAD

**Issue by PHMT-CW** [logic-ng/LogicNG#29]
Hi,

   Just wanted to start off by saying this is a great library to work with Boolean Formula's and solvers, keep up the good work. I've been developing an algorithm using the BDD implementation provided in the LogicNG library and encountered a couple of limiting factors for both appending to the BDD & finally converting the result of the "allSat" (byte[]) method back it's input Variables. 

First of all, the major issue I encountered during my use of the BDD implementation was that in order to use all the BDD functionality I am forced to use a BDDKernel class for the rest of the BDD implementation to function and therefore when some methods that I require for my algorithm to work are restricted due to being private I was forced to duplicate the entire BDDKernel source and have it extend the BDDKernel too. This resulted in duplication of all the variables found in the BDDKernel class which are all mostly private. This is not ideal and the major reason why I'm trying to reach out and request a simple fix that would help me use the BDDKernel without having to duplicate the code into my own source (including the BDDCache, BDDFactory, BDDPrime). 

If the method "makeNode" was changed from a private to protected modifier, this would correct my problem and opens up the possibility for basic extensions of the Kernel allowing developers to add to the node list and construct there own methods and custom functionality.

Secondly, I identified that the BDDFactory already contains a mapping by Variable Ordering to index which matches exactly to the byte[] output you receive from calling the allSat (1) method from the BDD, obviously because we are actually talking about the "level" the variable is on. If a simple method was providing in the factory to retrieve the Variable based on the index it would make everything much easier for developers to construct a result from the byte array.

```
/**
 * Returns a Variable object matching to the index according to the Variable Ordering.
 *
 * @param index - The expected index of the Variable
 * @return Variable - the Variable to which the specified index is mapped, otherwise null if the index contains no mapping
 */
public Variable getVariableByIndex(int index) {
    return this.idx2var.get(index);
}
```

Here is an example of how I converted a byte array into it a list of Variables:

```
return cutSet -> IntStreamEx.range(cutSet.length)
                    .filter(index -> cutSet[index] == 1)
                    .mapToObj(factory::getVariableByIndex)
                    .toImmutableList();
```

Thanks

**Comment by czengler** [comment]
Hi,

thank you very much for your kind words. In the upcoming release of LogicNG 2.0 which is currently under development and which is expected in June 2020, most of the `private` modifiers of LogicNG (including the ones you want) will be `protected` therefore allowing other developers to overwrite them in their own subclasses.

Also the additional method for retrieving variables by their index is no problem to include.

So the question is: is a productive release in June 2020 enough for your project, or would you require a release 1.6.3 - since we use the library quite heavily in many business critical applications, we always want to do proper releases of new functionality.

**Comment by PHMT-CW** [comment]
Thanks for your response, at this stage we can continue with our current implementation until the release in June. I understand that many business critical applications are using the library and you can now add our application to that list, it is significantly important that the result we get back from the library is correct and accurate (as a matter of safety & criticality) therefore we are happy to wait if you can assure everything is tested and correct. 

Thanks!

**Comment by PHMT-CW** [comment]
Thanks LogicNG Team, I updated to the latest version of the library and was able to refactor my code to how it needed to be. I was able to easily extend the Kernel and add my custom code without any issue. The adjustments to the BDDFactory and the introduction of BDDOperations was fine, it was easy to instantiate a new BDDOperations instance to help derive the 1 solutions. Keep up the good work.

One last thing that we discovered when exercising the BDD Kernel with a large Formula input was that the the code used to expand the size of the nodes array was hitting an IndexOutOfBoundsException because it didn't seem to grow big enough after resize. It then became necessary for us to scale the initial node array size to about 100 times the size of our known node count to avoid the condition where the node array would be resized. If the issue still persists in this version I'll open a new ticket to see if you guys can find a fix. Ideally if the node array could be replaced with a Collection implementation (i.e. a Map) it could grow more dynamically and maintain the requirement of finding nodes by a hash.
